pkg/topology/kademlia/internal/metrics/metrics.go

```diff
@@ -228,105 +250,82 @@ func (c *Collector) Record(addr swarm.Address, rop ...RecordOp) error {
 // be evaluated against the last seen time, which equals to the login time. If
 // the peer is logged out, then the session counters will reflect its last
-// session. The execution doesn't stop if some metric collection returns an
-// error, it rather continues and all the execution errors are returned together
-// with the successful metrics snapshots.
-func (c *Collector) Snapshot(t time.Time, addresses ...swarm.Address) (map[string]*Snapshot, error) {
-
-	var mErr error
+// session.
+func (c *Collector) Snapshot(t time.Time, addresses ...swarm.Address) map[string]*Snapshot {
 	snapshot := make(map[string]*Snapshot)
 
-	take := func(addr string) {
-
-		c.mu.RLock()
-		cs := c.counters[addr]
-		c.mu.RUnlock()
-
-		if cs == nil {
-			return
-		}
-
-		ls, err := cs.lastSeenTimestamp.Get()
-		if err != nil {
-			mErr = multierror.Append(mErr, fmt.Errorf("unable to take last seen snapshot for %q: %w", addr, err))
-		}
-		lastSeenTimestamp := int64(ls)
-
-		cn, err := cs.connTotalDuration.Get()
-		if err != nil {
-			mErr = multierror.Append(mErr, fmt.Errorf("unable to take connection duration snapshot for %q: %w", addr, err))
-		}
-		connTotalDuration := time.Duration(cn)
-
-		cs.Lock()
-		sessionConnDuration := cs.sessionConnDuration
-		if cs.loggedIn {
-			sessionConnDuration = t.Sub(time.Unix(0, lastSeenTimestamp))
-			connTotalDuration += sessionConnDuration
-		}
-
-		snapshot[addr] = &Snapshot{
-			LastSeenTimestamp:          lastSeenTimestamp,
-			SessionConnectionRetry:     cs.sessionConnRetry,
-			ConnectionTotalDuration:    connTotalDuration,
-			SessionConnectionDuration:  sessionConnDuration,
-			SessionConnectionDirection: cs.sessionConnDirection,
-		}
-		cs.Unlock()
-	}
-
 	for _, addr := range addresses {
-		take(addr.String())
+		val, ok := c.counters.Load(addr.ByteString())
+		if !ok {
+			continue
+		}
+		cs := val.(*Counters)
+		snapshot[addr.ByteString()] = cs.snapshot(t)
 	}
 
 	if len(addresses) == 0 {
-		for _, addr := range c.keys() {
-			take(addr)
-		}
+		c.counters.Range(func(key, val interface{}) bool {
+			cs := val.(*Counters)
+			snapshot[cs.peer.ByteString()] = cs.snapshot(t)
+			return true
+		})
 	}
 
-	return snapshot, mErr
+	return snapshot
 }
 
-func (c *Collector) keys() []string {
-
-	c.mu.RLock()
-	defer c.mu.RUnlock()
+// Inspect allows to inspect current snapshot for the given
+// peer address by executing the inspection function.
+func (c *Collector) Inspect(addr swarm.Address, fn func(ss *Snapshot)) {
+	snapshots := c.Snapshot(time.Now(), addr)
+	fn(snapshots[addr.ByteString()])
+}
 
-	keys := make([]string, 0, len(c.counters))
+// Flush sync the dirty in memory counters by flushing their values to the
+// underlying storage. If an address or a set of addresses is specified then
+// only counters related to them will be flushed, otherwise counters for all
+// peers will be flushed.
+func (c *Collector) Flush(addresses ...swarm.Address) error {
+	var mErr error
```

Comment:
Given that Counters.flush method creates a very small batch, I would go with suggestion from @acud to create a batch here and write all counters as a singe batch at the end of this function. This is not critical as current implementation is good enough and this optimization can be easily added if it is found that it is needed.